### PR TITLE
DOM

### PR DESCRIPTION
**Pull request by ka7eh**
How to test with ctsm_dev_env_template:

1. Clone this repo: https://github.com/MetOs-UiO/ctsm_dev_env_template
2. Create a `.env` file in the repo root with the following content:
  ```
  CTSM_REPO=https://github.com/MetOs-UiO/CTSM.git
  CTSM_TAG=kaveh/dom
  INPUTDATA=/mnt/fram/noresm/inputdata
  ```
3. Download the prepared ALP1 data from here: https://ns2806k.webs.sigma2.no/EMERALD/EMERALD_platform/ALP1.tar
4. Extract the data in `data` folder in the cloned repo
5. Run `./run_docker.sh` in the repo root, then run the following commands in the opened shell within docker container (keep the shell open after running the commands for later steps):
  - `/ctsm-api/scripts/setup_case.sh -n ALP1 -c 
2000_DATM%GSWP3v1_CLM51%BGC_SICE_SOCN_SGLC_SWAV`
  - `cd /ctsm-api/resources/cases/ALP1`
  - `./xmlchange STOP_OPTION=nyears`
  - `export CESMDATAROOT=/ctsm-api/resources/data/ALP1`
6. Add the following to `/ctsm-api/resources/cases/ALP1/user_nl_clm` from within the docker container, or `./cases/ALP1/user_nl_clm` from the repo root (if it's not already there):
  ```
  paramfile = '$CLM_USRDAT_DIR/clm50_params.c211112.nc'

 do_harvest = .false.
 check_dynpft_consistency = .false.
  ```
7. In the shell you started in step 5, run `./case.build` and then `case.submit`. If these are successful, you should see output in `ctsm-api/resources/cases/ALP1/archive/lnd/hist` from within the container or in `cases/ALP1/archive/lnd/hist` from the repo folder.

### CODE REVIEW

**Comment by devarajun** [comment]
I have performed a test run with this code for ALP1 site using NorESM_LSP and analyzed DOC fluxes: 5 year simulation (2000-2004) for the COMPSET: `2000_DATM%GSWP3v1_CLM50%FATES_SICE_SOCN_MOSART_SGLC_SWAV_SESP`. See figure below 
DOCprod_6p
the DOC flux values corresponding to docf_l1s1 (CEL_LIT_DOC_vr) and l2s1 (MET_LIT_DOC_vr) seems fine but comparing with other output variables GPP/NPP/HR will give some indication of their magnitudes. However, docf_s1s2 (ACT_SOM_DOC_S2_vr) and docf_s1s3 (PAS_SOM_DOC_vr) are zeros and to confirm this I think we need to run the ALP1 case with BGC mode instead of FATES (COMPSET `2000_DATM%GSWP3v1_CLM51%BGC_SICE_SOCN_SGLC_SWAV` ). 

Also TOTDOC variable do not have any values, this could be cumulative sum of DOC fluxes (CEL_LIT_DOC_vr, MET_LIT_DOC_vr, ACT_SOM_DOC_S2_vr and ACT_SOM_DOC_S3_vr ). This variable needs to be traced back on how this is calculated.

Also, need to run the long simulation (maybe 1000 years) for soil carbon spin-up to make sure ACT_SOM_DOC_S2_vr and PAS_SOM_DOC_vr are non-zero .

**Comment by devarajun** [comment]
Here is a python code to generate the figure given above :
```
import matplotlib.pyplot as plt
import xarray as xr

#list variables
varname = ['CEL_LIT_DOC_vr','LIG_LIT_DOC_vr','MET_LIT_DOC_vr','ACT_SOM_DOC_S2_vr','PAS_SOM_DOC_vr','SLO_SOM_DOC_S1_vr']

#define figure size
fig = plt.figure(figsize=[20,18])
plt.subplots_adjust(left=0.1,right=0.99,hspace=0.2,wspace=0.2)

i=0 #iteration index
for var in varname:
    print(var)
    i=i+1
    f1 = xr.open_dataset(var+'_5yr.nc')
    vardoc = f1[var]
    print('DOC variable',vardoc.shape)
    vardoc1d = vardoc.isel(levdcmp=0,lndgrid=0)
    #create figure
    ax = fig.add_subplot(3,2,i)
    ax.plot(vardoc1d)
    ax.set_ylabel(var)
    ax.set_xlabel('Months (2000-2004)')
fig.savefig('DOCprod_6p.png')

```

**Comment by devarajun** [comment]
Update: 
I performed the test run with the COMPSET: `2000_DATM%GSWP3v1_CLM51%BGC_SICE_SOCN_SGLC_SWAV`
and also noticed the doc fraction numbers were not updated (docf_l1s1=1.43e-4, docf_l2s1=7.143e-5, docf_s1s2=7.143e-4, docf_s1s3=1.43e-4). 

Below figure are the results after updating docf numbers in the code.
DOCprod_20yr

I figured out that the DOC flux variables corresponding to docf_s1s2 is ACT_SOM_DOC_S2_vr and docf_s1s3 is ACT_SOM_DOC_S3_vr . For the numbers, if we compare these variable results with HR variable, the numbers for DOC flux variables seems fine. 

I still see TOTDOC variable has no values and it needs to be traced back. This variable values could be cumulative sum of CEL_LIT_DOC_vr, MET_LIT_DOC_vr, ACT_SOM_DOC_S2_vr and ACT_SOM_DOC_S3_vr.

**Comment by devarajun** [comment]
Here is the commit that I have updated the code with docf* numbers:
[commit](https://github.com/MetOs-UiO/CTSM/commit/956a2e39b54dee42da1c064c9be67c14b7ee942e)

**Comment by ka7eh** [comment]
> Here is a python code to generate the figure given above :
> 
> ```
> import matplotlib.pyplot as plt
> import xarray as xr
> 
> #list variables
> varname = ['CEL_LIT_DOC_vr','LIG_LIT_DOC_vr','MET_LIT_DOC_vr','ACT_SOM_DOC_S2_vr','PAS_SOM_DOC_vr','SLO_SOM_DOC_S1_vr']
> 
> #define figure size
> fig = plt.figure(figsize=[20,18])
> plt.subplots_adjust(left=0.1,right=0.99,hspace=0.2,wspace=0.2)
> 
> i=0 #iteration index
> for var in varname:
>     print(var)
>     i=i+1
>     f1 = xr.open_dataset(var+'_5yr.nc')
>     vardoc = f1[var]
>     print('DOC variable',vardoc.shape)
>     vardoc1d = vardoc.isel(levdcmp=0,lndgrid=0)
>     #create figure
>     ax = fig.add_subplot(3,2,i)
>     ax.plot(vardoc1d)
>     ax.set_ylabel(var)
>     ax.set_xlabel('Months (2000-2004)')
> fig.savefig('DOCprod_6p.png')
> ```

The code here expects netcdf files in the working directory for the vars specified, like CEL_LIT_DOC_vr_5yr.nc. Can you add something on how to get those files from the output?

File: src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90 [hunk 1]
```diff
@@ -302,6 +381,11 @@ subroutine init_decompcascade_bgc(bounds, soilbiogeochem_state_inst, soilstate_i
             f_s1s3(c,j) = .004_r8 / (1._r8 - t)
             rf_s1s2(c,j) = t
             rf_s1s3(c,j) = t
+            ! TODO check [kaveh] why do we start with 0?
+            docf_s1s2(c,j) = 0.0_r8
+            docf_s1s3(c,j) = 0.0_r8
+            docf_s1s2(c,1) = 0.0007143_r8
```
Comment (ka7eh) [inline src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90]: @devarajun where are these numbers coming from? Do they need to be hard-coded?

File: src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90 [hunk 2]
```diff
@@ -451,58 +535,76 @@ subroutine init_decompcascade_bgc(bounds, soilbiogeochem_state_inst, soilstate_i
       !----------------  list of transitions and their time-independent coefficients  ---------------!
       i_l1s1 = 1
       decomp_cascade_con%cascade_step_name(i_l1s1) = 'L1S1'
+      docf_decomp_cascade(bounds%begc:bounds%endc,1:nlevdecomp,i_l1s1) = 0.000143_r8
```
Comment (ka7eh) [inline src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90]: Same question as above: where are these coming from? And should they be hard-coded?

File: src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90 (hunk 1 above)
Comment (devarajun) [reply src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90]: These numbers should be read from netcdf paramafile in the inputdata. For now, I put the numbers in the code.

File: src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90 (hunk 2 above)
Comment (devarajun) [reply src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90]: This number also should be read from netcdf paramfile, if we follow `rf_11s1` format.

**Comment by devarajun** [comment]
Here is a python code to generate the figure given above :

```
import matplotlib.pyplot as plt
import xarray as xr

#list variables
varname = ['CEL_LIT_DOC_vr','MET_LIT_DOC_vr','ACT_SOM_DOC_S2_vr','ACT_SOM_DOC_S3_vr','NPP','HR']

#define figure size
fig = plt.figure(figsize=[20,18])
plt.subplots_adjust(left=0.1,right=0.99,hspace=0.2,wspace=0.2)

i=0 #iteration index
for var in varname:
    print(var)
    i=i+1
    f1 = xr.open_dataset(var+'_5yr.nc',decode_times=False)
    vardoc = f1[var]
    print('DOC variable',vardoc.shape)
    if (var=='NPP' or var=='HR'):
        vardoc1d = vardoc.isel(lndgrid=0)
    else:
        vardoc1d = vardoc.isel(levdcmp=0,lndgrid=0)
    #create figure
    ax = fig.add_subplot(3,2,i)
    ax.plot(vardoc1d)
    ax.set_ylabel(var+'('+vardoc.units+')')
    ax.set_xlabel('Months (2000-2019)')
fig.savefig('DOCprod_5yr.png')

  
```

The code here expects netcdf files in the working directory for the vars specified, like CEL_LIT_DOC_vr_5yr.nc. Can you add something on how to get those files from the output?

Below is a simple bash script that extracts required variable from multiple netcdf files into 1 netcdf file:

```
#!/bin/bash
#need to install nco software eg. conda install -c conda-forge nco

#extract variables from multiple netcdf files into one netcdf file
dataarchive=/home/devaraju/Documents/UiO/ctsm_dev_env_template/cases/ALP1/archive/lnd/hist
varname=("CEL_LIT_DOC_vr" "MET_LIT_DOC_vr" "ACT_SOM_DOC_S2_vr" "ACT_SOM_DOC_S3_vr" "NPP" "HR")
for var in ${varname[@]};do
    ncrcat -v $var $dataarchive/*ALP1.*.nc ${var}_5yr.nc
done
```

**Pull request closed without merging**